source/stdx/collections/rcstring.d

```diff
@@ -0,0 +1,244 @@
+     * $(REF, GCAllocator, std,experimental,allocator,gc_allocator).
+     *
+     * Params:
+     *      bytes = a variable number of bytes, either in the form of a
+     *               list or as a built-in array
+     *
+     * Complexity: $(BIGOH m), where `m` is the number of bytes.
+     */
+    this(this Q)(ubyte[] bytes...)
+    {
+        static if (is(Q == immutable) || is(Q == const))
+        {
+            this(processAllocatorObject(), bytes);
+        }
+        else
+        {
+            this(threadAllocatorObject(), bytes);
+        }
+    }
+
+    ///
+    @safe unittest
+    {
+        // Create a list from a list of bytes
+        {
+            auto a = RCString('1', '2', '3');
+        }
+        // Create a list from an array of bytes
+        {
+            auto a = RCString(['1', '2', '3']);
+        }
+    }
+
+    /**
+     * Constructs a qualified rcstring out of a number of bytes
+     * that will use the provided allocator object.
+     * For `immutable` objects, a `RCISharedAllocator` must be supplied.
+     *
+     * Params:
+     *      allocator = a $(REF RCIAllocator, std,experimental,allocator) or
+     *                  $(REF RCISharedAllocator, std,experimental,allocator)
+     *                  allocator object
+     *      bytes = a variable number of bytes, either in the form of a
+     *               list or as a built-in RCString
+     *
+     * Complexity: $(BIGOH m), where `m` is the number of bytes.
+     */
+    this(A, this Q)(A allocator, ubyte[] bytes...)
+    if (!is(Q == shared)
+        && (is(A == RCISharedAllocator) || !is(Q == immutable))
+        && (is(A == RCIAllocator) || is(A == RCISharedAllocator)))
+    {
+        debug(CollectionRCString)
+        {
+            writefln("RCString.ctor: begin");
+            scope(exit) writefln("RCString.ctor: end");
+        }
+        static if (is(Q == immutable) || is(Q == const))
+        {
+            _allocator = immutable AllocatorHandler(allocator);
+        }
+        else
+        {
+            setAllocator(allocator);
+        }
+        _support = typeof(_support)(allocator, bytes);
+    }
+
+    ///
+    @safe unittest
+    {
+        import std.experimental.allocator : theAllocator, processAllocator;
+
+        // Create a list from a list of ints
+        {
+            auto a = RCString(theAllocator, '1', '2', '3');
+        }
+        // Create a list from an array of ints
+        {
+            auto a = RCString(theAllocator, ['1', '2', '3']);
+        }
+    }
+
+    //char front() const
+    //{
+        //auto c = _support.front;
+        //return c;
+    //}
+
+    //void popFront()
+    //{
+        //_support.popFront;
+    //}
+
+    bool empty() const
+    {
+        return _support.empty;
+    }
+
+    @trusted
+    auto by(T)()
+    if (is(T == char) || is(T == wchar) || is(T == dchar))
+    {
+        Array!char tmp = *cast(Array!char *)(&_support);
```

Comment:
Why do we need this casting here and not use support directly?
Also mind the pointer D style - it's char*